core/storage/blog_statistics/gae_models.py

```diff
@@ -0,0 +1,999 @@
+    read.
+    """
+
+    # ID of blog post currently being read.
+    blog_post_id = datastore_services.StringProperty(
+        indexed=True, required=True)
+    # User ID of the author of the blog post.
+    author_id = datastore_services.StringProperty(indexed=True, required=True)
+
+    @staticmethod
+    def get_deletion_policy() -> base_models.DELETION_POLICY:
+        """Model contains data corresponding to a user and should be deleted."""
+        return base_models.DELETION_POLICY.DELETE
+
+    @classmethod
+    def get_new_event_entity_id(cls, blog_post_id: str) -> str:
+        """Generates a unique id for the event model of the form
+        '[timestamp]:[blog_post_id]:[random_hash]'.
+        """
+        timestamp = datetime.datetime.utcnow()
+
+        # To avoid collision between events occuring at the same date and time,
+        # a random hash is appended to the event model id.
+        for _ in range(base_models.MAX_RETRIES):
+            random_hash = utils.convert_to_hash(
+                str(utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+        return cls.get_new_id('%s:%s:%s' % (
+            utils.get_time_in_millisecs(timestamp),
+            blog_post_id,
+            random_hash))
+
+    @classmethod
+    def create(
+            cls,
+            blog_post_id: str,
+            author_id: str,
+    ) -> str:
+        """Creates a new blog post read event entry."""
+        entity_id = cls.get_new_event_entity_id(blog_post_id)
+        event_entity = cls(
+            id=entity_id,
+            blog_post_id=blog_post_id,
+            author_id=author_id)
+        event_entity.update_timestamps()
+        event_entity.put()
+        return entity_id
+
+    @staticmethod
+    def get_model_association_to_user() -> (
+        base_models.MODEL_ASSOCIATION_TO_USER
+    ):
+        """Model does not contain neccessary information for user for
+        takeout.
+        """
+        return base_models.MODEL_ASSOCIATION_TO_USER.NOT_CORRESPONDING_TO_USER
+
+    @classmethod
+    def get_export_policy(cls) -> Dict[str, base_models.EXPORT_POLICY]:
+        """Model doesn't contain any data directly corresponding to a user that
+        should be exported.
+        """
+        return dict(super(cls, cls).get_export_policy(), **{
+            'blog_post_id': base_models.EXPORT_POLICY.NOT_APPLICABLE,
+            'author_id': base_models.EXPORT_POLICY.NOT_APPLICABLE
+        })
+
+    @classmethod
+    def has_reference_to_user_id(cls, user_id: str) -> bool:
+        """Check whether EventModel references user.
+
+        Args:
+            user_id: str. The ID of the user whose data should be checked.
+
+        Returns:
+            bool. Whether any models refer to the given user ID.
+        """
+        return cls.query(
+            cls.author_id == user_id
+        ).get(keys_only=True) is not None
+
+    @classmethod
+    def get_all_by_author(
```

Comment:
What uses this? Not sure it will scale.

Normally for raw event log models we just write them to the datastore, but we only use them when we need to regenerate summary data in Beam jobs. There are lots of event models so fetching them directly will time out. If you want summary stats then update the summary models at time of writing, don't use this.

Reply:
Removed form here and other classes.

Reply:
Done

Reply:
Done